### ISSUE TITLE
Why does data loading code use so much RAM?

### ISSUE THREAD

**Issue by JackKelly** [openclimatefix/predict_pv_yield_2#22]
Uses > 50 GB RAM when using multiple workers.

Things to try:

- [ ] Try without `PVDataLoader` - are we using lots of RAM because the PV data is getting copied lots of times?
- [ ] Only load NWPs for the land area of the UK (not the oceans)

**Comment by JackKelly** [comment]
The NWPDataLoader is definitely causing issues.  If we only load the first 4 `steps` then that reduces memory usage a lot.

**Comment by JackKelly** [comment]
NWPDataInMem.get_sample takes about 70 ms per sample.  So with 8 samples per batch, it takes over half a second.  That's probably the issue.

**Comment by JackKelly** [comment]
Yup.  The interpolation (even linear) takes a while.  Replacing the linear interpolation for `ffill` decreases the runtime of `NWPDataInMem.get_sample` from 70 ms to 15 ms.  And increases the training speed from about 5 it/s to 15 it/s.

**Comment by JackKelly** [comment]
We're probably using lots of ram for the same reason as it's running a little slow (#24):  We're resampling to 5 minutes the _entire_ NWP image!  See https://github.com/openclimatefix/predict_pv_yield_2/issues/24#issuecomment-841419551 for ideas on how to fix this

**Comment by JackKelly** [comment]
Tried ensuring threads go out of scope, but that doesn't help:

```python
    def __iter__(self):
        """Asynchronously loads next data from disk while sampling from data in memory."""
        # Load initial data.
        with futures.ThreadPoolExecutor() as executor:
            future_data_loaders, selected_zarr_chunk_sequences = self._async_load(executor)
            in_mem_data_from_asyc_loaders = [
                [future.result() for future in future_data_per_chunk]
                for future_data_per_chunk in future_data_loaders]

        for _ in range(self.n_disk_loads_per_epoch):
            with futures.ThreadPoolExecutor() as executor:
                future_data_loaders, next_selected_zarr_chunk_sequences = self._async_load(executor)
                # Pick a random number of samples to sample from one disk load.
                # We do this so that the worker processes don't all try to load from disk at the same time.
                n_samples = self.rng.integers(self.min_n_samples_per_disk_load, self.max_n_samples_per_disk_load)
                for _ in range(n_samples):
                    # Randomly select chunk_i, start_time and end_time.
                    chunk_i = self.rng.integers(0, self.n_zarr_chunk_sequences_to_load_at_once)
                    segment = selected_zarr_chunk_sequences[chunk_i]
                    start_time = self._get_random_start_time(segment)
                    # Subtract 1 timestep (5 minutes) from end_time because .loc[start:end] includes the start and end,
                    # so will give us 1 extra timestep!
                    end_time = start_time + self.total_seq_timedelta - pd.Timedelta('5 minutes')
                    datetime_index = pd.date_range(start_time, end_time, freq='5T')
                    t0 = datetime_index[self.history_len - 0]

                    # Initialise Sample.  Sample will be updated with more data in the next code block.
                    sample = Sample(
                        datetime_index=datetime_index,
                        history_len=self.history_len,
                        forecast_len=self.forecast_len)
                    
                    # Collect data from AsyncDataLoaders.
                    # TODO: Could use multiple threads to load data from all the async_loaders in parallel.
                    for in_mem_data_per_chunk in in_mem_data_from_asyc_loaders:
                        in_mem_data = in_mem_data_per_chunk[chunk_i]
                        try:
                            new_sample = in_mem_data.get_sample(start_time, end_time, t0)
                        except:
                            print(f'Failed get_sample.  segment={segment}')
                            raise
                        sample.update(new_sample)

                    # Transform.
                    if self.transform:
                        try:
                            sample = self.transform(sample)
                        except BadData as e:
                            # print(e)
                            continue

                    yield sample

                # Prepare for next iteration.
                selected_zarr_chunk_sequences = next_selected_zarr_chunk_sequences
                in_mem_data_from_asyc_loaders = [
                    [future.result() for future in future_data_per_chunk]
                    for future_data_per_chunk in future_data_loaders]
```

**Comment by JackKelly** [comment]
There are two problems:

1. The code uses lots of RAM, even at the start of training
2. The memory usage climbs slowly during training.  I think I've fixed this by using `xr.DataArray.compute` instead of `xr.DataArray.load`.  The latter modifies the original data array, and returns the new one.

**Comment by JackKelly** [comment]
Very weirdly, changing the data type of PV data and NWP data doesn't seem to change the memory usage.

**Comment by JackKelly** [comment]
Also, what happens if we don't use multiple threads in the loader process?!?!?

**Comment by JackKelly** [comment]
pinning memory doesn't seem to affect memory usage.

**Comment by JackKelly** [comment]
reducing `n_zarr_chunk_sequences_to_load_at_once` from 8 to 4 does almost halve memory usage (as expected)

**Comment by JackKelly** [comment]
Limiting the number of threads doesn't help:  `with futures.ThreadPoolExecutor(max_workers = len(self.async_data_loaders) * self.n_zarr_chunk_sequences_to_load_at_once) as executor:`

**Comment by JackKelly** [comment]
Maybe the code shouldn't be explicitly triggering multiple threads; instead maybe we should be using the [`async` functionality of dask](https://examples.dask.org/applications/async-await.html)?  And investigating what dask is doing under the hood with its viewer thing.

**Comment by JackKelly** [comment]
This is fixed by work done in #27
